### PR TITLE
Separate reference domain dimension and nodal coordinate dimension

### PR DESCRIPTION
**Pull request by yijiangh**
### Summary

This PR tries to split the `dim` type parameter to `ξdim` and `xdim`. Before, the dimension of the reference domain (`\xi`) is coupled with the dimension of the nodal coordinate domain.

This PR starts with changing `Values{dim,T,refshape}` to `Values{ξdim,xdim,T,refshape}`, and other changes are made to make things compatible and non-breaking. The `dim` in function argument are renamed to `ξdim` or `xdim` to clarify the context.
This PR is particularly useful for modeling non-standard cell in JuAFEM, e.g. truss elements (1D `Line` cell with nodes in 2D or 3D).

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [x] Other (e.g. doc update, configuration, etc)

### Checklist

- [x] I ran all tests on my computer and it's all green (i.e. `] test`).
   - All local tests (julia 1.5.0 on Windows) passed except the ones involve file sha comparison. I'm not sure if the CI tests would pass or not 
```julia
Test Summary:         | Fail  Total
Grid, DofHandler, vtk |   16     16
```
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/FEValues/cell_values.jl [hunk 1]
```diff
@@ -78,54 +87,61 @@ function CellScalarValues(::Type{T}, quad_rule::QuadratureRule{dim,shape}, func_
...
-    CellScalarValues{dim,T,shape}(N, dNdx, dNdξ, detJdV, M, dMdξ, quad_rule.weights)
+    CellScalarValues{ξdim,xdim,T,shape}(N, dNdx, dNdξ, detJdV, M, dMdξ, quad_rule.weights)
 end
 
 # CellVectorValues
-struct CellVectorValues{dim,T<:Real,refshape<:AbstractRefShape,M} <: CellValues{dim,T,refshape}
-    N::Matrix{Vec{dim,T}}
-    dNdx::Matrix{Tensor{2,dim,T,M}}
-    dNdξ::Matrix{Tensor{2,dim,T,M}}
+# ? Vector dimension is assumed to be xdim ?
```
Comment (yijiangh) [inline src/FEValues/cell_values.jl:94]: Are we assuming the vector dimension is `xdim`? I think this is related to a [comment ](https://github.com/KristofferC/JuAFEM.jl/issues/193#issuecomment-502247133) made in #193.

File: src/FEValues/cell_values.jl [hunk 2]
```diff
@@ -78,54 +87,61 @@ function CellScalarValues(::Type{T}, quad_rule::QuadratureRule{dim,shape}, func_
...
 
     # Function interpolation
-    n_func_basefuncs = getnbasefunctions(func_interpol) * dim
-    N    = fill(zero(Vec{dim,T})      * T(NaN), n_func_basefuncs, n_qpoints)
-    dNdx = fill(zero(Tensor{2,dim,T}) * T(NaN), n_func_basefuncs, n_qpoints)
-    dNdξ = fill(zero(Tensor{2,dim,T}) * T(NaN), n_func_basefuncs, n_qpoints)
+    n_func_basefuncs = getnbasefunctions(func_interpol) * xdim
+    N    = fill(zero(Vec{xdim,T})      * T(NaN), n_func_basefuncs, n_qpoints)
+    dNdx = fill(zero(Tensor{2,xdim,T}) * T(NaN), n_func_basefuncs, n_qpoints)
+    # TODO: each entry should be a second order tensor with shape xdim x ξdim
```
Comment (yijiangh) [inline src/FEValues/cell_values.jl:125]: See comment: https://github.com/KristofferC/JuAFEM.jl/pull/295/files#r497813333

File: src/FEValues/face_values.jl [hunk 3]
```diff
@@ -93,18 +100,20 @@ function FaceScalarValues(::Type{T}, quad_rule::QuadratureRule{dim_qr,shape}, fu
...
-    FaceScalarValues{dim,T,shape}(N, dNdx, dNdξ, detJdV, normals, M, dMdξ, quad_rule.weights, ScalarWrapper(0))
+    FaceScalarValues{ξdim,xdim,T,shape}(N, dNdx, dNdξ, detJdV, normals, M, dMdξ, quad_rule.weights, ScalarWrapper(0))
 end
 
 # FaceVectorValues
-struct FaceVectorValues{dim,T<:Real,refshape<:AbstractRefShape,M} <: FaceValues{dim,T,refshape}
-    N::Array{Vec{dim,T},3}
-    dNdx::Array{Tensor{2,dim,T,M},3}
-    dNdξ::Array{Tensor{2,dim,T,M},3}
+# ? Vector dimension is assumed to be xdim ?
```
Comment (yijiangh) [inline src/FEValues/face_values.jl:107]: See comment: https://github.com/KristofferC/JuAFEM.jl/pull/295/files#r497813333

File: src/FEValues/face_values.jl [hunk 4]
```diff
@@ -40,14 +40,15 @@ For a scalar field, the `FaceScalarValues` type should be used. For vector field
 FaceValues
 
 # FaceScalarValues
-struct FaceScalarValues{dim,T<:Real,refshape<:AbstractRefShape} <: FaceValues{dim,T,refshape}
+# ? why 3 here? what if we have a hex cell?
```
Comment (yijiangh) [inline src/FEValues/face_values.jl:43]: Question here.

**Comment by lijas** [comment]
I agree that there should be a way to model beams/shells/trusses etc in JuAFEM. I dont think the dofhandlers can currently correctly distribute dofs for shell elements if one wants 2nd order interpolations. This is due to how faces/edges are defined differently in 2d/3d interpolations. I did however try to solve this a while back in this branch https://github.com/lijas/JuAFEM.jl/tree/line2

My branch makes it possible to do something like this:
```
grid = Grid(Line3d, ...) #Grid in space-dim=3, but with line elements

#3d line cells with 2nd order 1d interpolation
dh = DofHandler(grid)
push!(dh, :u, 3, Lagrange{1,RefCube,2}()) 
push!(dh, :θ, 3, Lagrange{1,RefCube,2}()) 
close!(dh)
```

I dont think it is needed to rename every dim to xdim. Lets keeps dim for the space-dimension.
One has to rewrite the cell-values and add the parameter dim, which you seem to have done.

**Comment by yijiangh** [comment]
> `push!(dh, :θ, 3, Lagrange{1,RefCube,2}()) `

This is great! May I ask why the number of faces `M` equal to 1 and 0 for `Line2d` and `Line3d` in your branch [here](https://github.com/lijas/JuAFEM.jl/blob/line2/src/Grid/grid.jl#L27-L28)?

> I dont think it is needed to rename every dim to xdim. Lets keeps dim for the space-dimension.

Hmm, I only changed `dim` whenever there is a confusion. We are not using named arguments almost anywhere in `JuAFEM` so I think this should be okay?

**Comment by yijiangh** [comment]
@lijas Also, do you have a working example for shell elements on your branch?

**Comment by lijas** [comment]
A line element in 3d has no faces, it has only one "edge".
A line element in 2d has only one face(and no edges, because edges only exist in 3d). Maybe one could argue that there is two faces in 2d (one on each side). 

Maybe these definitions for faces/edges are not always important, but when I have written generic contact algorithms for FE problems, I have noticed that it has been helpful. I also think it is the most correct.

I have no working shell example unfortunately. I just wrote some test to see that the dof distributions is correct for shells.

**Comment by fredrikekre** [comment]
Superseded by #694 + #651 (and some other supporting PRs). Thanks @yijiangh for the PR regardless!

**Pull request closed without merging**